pkg/controller/certificate-shim/controller.go

```diff
@@ -38,15 +42,71 @@ import (
 )
 
 const (
-	// ControllerName is the name of the ingress-shim controller.
-	ControllerName = "ingress-shim"
+	// IngressShimControllerName is the name of the ingress-shim controller.
+	IngressShimControllerName = "ingress-shim"
+
+	// GatewayShimControllerName is the name of the gateway-shim controller.
+	GatewayShimControllerName = "gateway-shim"
 )
 
 type defaults struct {
 	autoCertificateAnnotations          []string
 	issuerName, issuerKind, issuerGroup string
 }
 
+type ingressShim struct {
+	controller
+}
+
+func (i *ingressShim) Register(ctx *controllerpkg.Context) (workqueue.RateLimitingInterface, []cache.InformerSynced, error) {
+	// construct a new named logger to be reused throughout the controller
+	i.log = logf.FromContext(ctx.RootContext, IngressShimControllerName)
+
+	// create a queue used to queue up items to be processed
+	i.queue = workqueue.NewNamedRateLimitingQueue(controllerpkg.DefaultItemBasedRateLimiter(), IngressShimControllerName)
+
+	ingressInformer := ctx.KubeSharedInformerFactory.Networking().V1beta1().Ingresses()
+	i.objectInformer = ingressInformer.Informer()
+
+	i.objectLister = &internalIngressLister{ingressInformer.Lister()}
+
+	return i.sharedRegister(ctx)
+}
+
+type gatewayShim struct {
+	controller
+}
+
+func (g *gatewayShim) Register(ctx *controllerpkg.Context) (workqueue.RateLimitingInterface, []cache.InformerSynced, error) {
+	// construct a new named logger to be reused throughout the controller
+	g.log = logf.FromContext(ctx.RootContext, GatewayShimControllerName)
+
+	g.queue = workqueue.NewNamedRateLimitingQueue(controllerpkg.DefaultItemBasedRateLimiter(), GatewayShimControllerName)
+
+	// Discover if the gateway api is available
```

Comment:
nit: explain the why, not the "what"

```suggestion
	// The apiserver may not have the Gateway API CRDs installed, and we want to disable
	// the Gateway API support (shim and HTTP01 solver) when it is not found.
```

Reply:
👍  I will go through and add better comments. The gateway-shim controller is disabled by default, if it is explicitly enabled with the command line flag, but the gateway API CRDs have not been installed cert-manager will refuse to start.